src/singer_clojure/catalog.clj

```diff
@@ -0,0 +1,120 @@
+(ns singer-clojure.catalog)
+
+(defn get-selected-streams
+  [catalog]
+  "Takes a deserialized catalog and returns a sequence of the selected
+  tap-stream-ids"
+  (filter #(get-in catalog [% "metadata" "selected"]) (keys catalog)))
+
+;;
+;;
+;; SERIALIZATION
+;;
+;;
+(defn serialize-stream-metadata-property
+  [[stream-metadata-property-name stream-metadata-property-metadata :as stream-metadata-property]]
+  {"metadata"   stream-metadata-property-metadata
+   "breadcrumb" ["properties" stream-metadata-property-name]})
+
+(defn serialize-stream-metadata-properties
+  [stream-metadata-properties]
+  (let [properties (stream-metadata-properties "properties")]
+    (concat [{"metadata" (dissoc stream-metadata-properties "properties")
+              "breadcrumb" []}]
+            (map serialize-stream-metadata-property properties))))
+
+(defn serialize-stream-metadata
+  [{:keys [metadata] :as stream}]
+  (update stream "metadata" serialize-stream-metadata-properties))
+
+(defn serialize-metadata
+  [catalog]
+  (update catalog "streams" (partial map serialize-stream-metadata)))
+
+(defn serialize-stream-schema-property
+  [[k v]]
+  (if (nil? v)
+    [k {}]
+    [k v]))
+
+(defn serialize-stream-schema-properties
+  [stream-schema-properties]
+  (into {} (map serialize-stream-schema-property
+                stream-schema-properties)))
+
+(defn serialize-stream-schema
+  [stream-schema]
+  (update stream-schema
+          "properties"
+          serialize-stream-schema-properties))
+
+(defn serialize-stream
+  [stream-catalog-entry]
+  (update stream-catalog-entry "schema"
+          serialize-stream-schema))
+
+(defn serialize-streams
+  [catalog]
+  (assoc {}
+         "streams"
+         (map serialize-stream (vals catalog))))
+
+(defn serialize-catalog
```

Comment:
Is it the case that this function and the opposing deserialize are the only real "public" functions in this namespace? I think noting the other functions with `defn-` would be useful in denoting that they are "private" and not really part of the API that should be used.

Reply:
Yeah, I forgot about marking private functions with `defn-` -- These are updated now